### DISCUSSION TITLE
why Newtonsoft.Json too big when build native aot

### DISCUSSION THREAD

**Discussion by NCLnclNCL** [dotnet/runtime#88601] [General] [answered]
if i build with System.Text.Json it is 6 mb but it i build with Newtonsoft.Json it is 16 mb, both have the same same size of dll files

**Comment by ghost** [comment]
Tagging subscribers to this area: @dotnet/area-system-text-json, @gregsdennis
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.
<details>
<summary>Issue Details</summary>
<hr />

if i build with System.Text.Json it is 6 mb but it i build with Newtonsoft.Json it is 16 mb, both have the same size of dll files


<table>
  <tr>
    <th align="left">Author:</th>
    <td>NCLnclNCL</td>
  </tr>
  <tr>
    <th align="left">Assignees:</th>
    <td>-</td>
  </tr>
  <tr>
    <th align="left">Labels:</th>
    <td>

`area-System.Text.Json`

</td>
  </tr>
  <tr>
    <th align="left">Milestone:</th>
    <td>-</td>
  </tr>
</table>
</details>

**Comment by hez2010** [comment, marked as answer]
Newtonsoft.Json doesn't have NativeAOT support at all, so it instead rooting the whole assembly regardless of whether a type will be used or not.
While System.Text.Json has the support if you are using source generator, so only the necessary part of code will be kept in the binary.

**Reply by lindexi** [reply]
Before Newtonsoft.Json: 3MB

After add Newtonsoft.Json: 15MB

Test code: https://github.com/dotnet-campus/dotnetCampus.Ipc/commit/178366b4fd1fb1ad3761380960ac721b68db92e8

**Reply by MichalStrehovsky** [reply]
> Before Newtonsoft.Json: 3MB
> 
> After add Newtonsoft.Json: 15MB
> 
> Test code: [dotnet-campus/dotnetCampus.Ipc@178366b](https://github.com/dotnet-campus/dotnetCampus.Ipc/commit/178366b4fd1fb1ad3761380960ac721b68db92e8)

You can use [sizoscope](https://github.com/MichalStrehovsky/sizoscope) to diff the before and after binaries. But last time I looked into it, Newtonsoft json brought dependencies such as support for the `dynamic` keyword, System.Linq.Expressions, etc. There's a lot of untrimmable code within Newtonsoft Json, but also it depends on the whole world.

**Reply by lindexi** [reply]
@MichalStrehovsky Thank you very much for your power tool.

<img width="1306" height="789" alt="image" src="https://github.com/user-attachments/assets/74756f94-d786-41a9-8df6-613dd895fd8e" />
